### PR TITLE
No moar cursors

### PR DESCRIPTION
**Pull request by flintinatux**
hermes

It turns out that PG cursors aren't all they're cracked up to be, especially in combination with PG functions.  According to [the docs](https://www.postgresql.org/docs/12/plpgsql-control-structures.html):

> The current implementation of RETURN NEXT and RETURN QUERY stores the entire result set before returning from the function, as discussed above. That means that if a PL/pgSQL function produces a very large result set, performance might be poor: data will be written to disk to avoid memory exhaustion, but the function itself will not return until the entire result set has been generated. A future version of PL/pgSQL might allow users to define set-returning functions that do not have this limitation. Currently, the point at which data begins being written to disk is controlled by the work_mem configuration variable. Administrators who have sufficient memory to store larger result sets in memory should consider increasing this parameter.

Since Message DB is basically all PG server functions, the functions buffer the result set before returning, which means wrapping them in a cursor adds transactions, state, and increased memory usage with zero benefit.  So this PR drops back to nice, clean pagination.

Some added benefits:
1.  The `batchSize` option now has real meaning.
2.  Connections are released back to the pool sooner.
3.  One less dependency.

## to review:
- [ ] Read the updated docs.
- [ ] Wait for the green.
- [ ] Find a squirrel shaking his head at cursors.

### CODE REVIEW

File: src/db/getCategoryMessages.js [hunk 1]
```diff
@@ -1,26 +1,58 @@
...
+        debug('category messages found: %o', { category, count: rows.length })
+        push(null, rows)
+        next()
+      } else {
+        debug('end of category: %o', { category })
+        push(null, _.nil)
+      }
+    }
+
+    query(sql, vals).then(send, push)
```
Comment (mgreystone) [inline src/db/getCategoryMessages.js:45]: `.catch(push)`?

File: src/db/getStreamMessages.js [hunk 2]
```diff
@@ -1,23 +1,52 @@
...
+        debug('stream messages found: %o', { streamName, count: rows.length })
+        push(null, rows)
+        next()
+      } else {
+        debug('end of stream: %o', { streamName })
+        push(null, _.nil)
+      }
+    }
+
+    query(sql, vals).then(send, push)
```
Comment (mgreystone) [inline src/db/getStreamMessages.js:39]: `.catch(push)`?

File: src/db/getCategoryMessages.js (hunk 1 above)
Comment (flintinatux) [reply src/db/getCategoryMessages.js:45]: Is that not equivalent to supplying `push` as the second argument to `.then()`?

Comment (mgreystone) [reply src/db/getCategoryMessages.js:45]: It sure is. I forget that `then` accepts two args. 😞

Comment (flintinatux) [reply src/db/getCategoryMessages.js]: Actually... it's not.  I tested in a repl.  If you `first().then(second, fail)`, then `fail` is not executed if `second`, er... fails.  So I've swapped it for `.catch()` instead.

**Comment by flintinatux** [comment]
squirrel stamp

**Pull request merged**